### PR TITLE
feat(ide-completion): extra sugar auto-completion `async fn ...` in `impl trait` for `async fn in trait` that's defined in desugar form

### PR DESCRIPTION
**Pull request by hyf0**
Solves #17719.

---

Preview

image

image

image

### CODE REVIEW

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -2181,6 +2181,53 @@ impl Function {
...
+            return false;
+        };
+
+        let Some(sync_trait_id) =
+            db.lang_item(self.ty(db).env.krate, LangItem::Sync).and_then(|t| t.as_trait())
+        else {
+            return false;
+        };
+
+        // TODO: There's no `LangItem::Send`. How do we get the id of `Send` trait?
```
Comment (hyf0) [inline crates/hir/src/lib.rs:2210]: Help wanted here🙋

File: crates/ide-completion/src/completions/item_list/trait_impl.rs [hunk 2]
```diff
@@ -210,7 +210,7 @@ fn add_function_impl(
                 ast::AssocItem::Fn(func) => func,
                 _ => unreachable!(),
             };
-
+            // TODO: need `function_decl` that unwraps future in the return type
```
Comment (hyf0) [inline crates/ide-completion/src/completions/item_list/trait_impl.rs:213]: Help wanted here🙋

File: crates/hir/src/lib.rs (hunk 1 above)
Comment (Veykril) [reply crates/hir/src/lib.rs:2210]: Huh that is very surprising. Goes to show that the Send/Sync stuff really is more a library level concern than a language one, interesting.

I think we can skip looking for those two all together and instead check whether the impl traits (aside from the principal `Future` one) are auto traits or not. That is more general, as the relevant part about those traits being leaked is them being auto traits, not them being `Send` or `Sync`

File: crates/ide-completion/src/completions/item_list/trait_impl.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide-completion/src/completions/item_list/trait_impl.rs:213]: You'll have to manually traverse the AST here and stitch the edit together, that is something like `transformed_fn.ret_type()` to fetch the return type, then you can match on that to check if its an `ImplTrait` one etc

**Comment by Veykril** [comment]
Finished up the work here, hope you don't mind :)

@bors r+

**Comment by hyf0** [comment]
Not at all. Thank you very much. Finally no more manual async anymore :).

**Pull request merged**